### PR TITLE
[Model] Add ColPali late interaction model for multi-modal retrieval

### PR DESCRIPTION
**Pull request by Kaonael**
Add support for ColPali (ColBERT-style late interaction on PaliGemma backbone) for multi-vector document retrieval and reranking.

Target model: vidore/colpali-v1.3-hf
Reference: https://arxiv.org/abs/2407.01449

## Purpose
Add support for https://arxiv.org/abs/2407.01449 — a ColBERT-style late interaction model built on PaliGemma backbone (SigLIP + Gemma) for multi-vector document retrieval and reranking. 
Target model: vidore/colpali-v1.3-hf.
                     
- Model implementation extending PaliGemmaForConditionalGeneration with a linear projection head for per-token embeddings                                                          
- Custom ColPaliConfig extending PaliGemmaConfig (no trust_remote_code needed)
- Registration in model registry, config registry, and test model registry                                                                                                         
- Multimodal processor handling for image+text pooling inputs                                                                                                                      
- Weight loading with support for multiple checkpoint naming conventions (HF transformers, colpali-engine)                                                                         
- Updated docs/models/supported_models.md   

## Test Plan
```bash
uv run vllm serve vidore/colpali-v1.3-hf --runner pooling
```

```bash
# Tokenize
curl -s -X POST http://0.0.0.0:8000/tokenize \
  -H "Content-Type: application/json" \
  -d '{"prompt": "What is machine learning?"}'
# Detokenize
curl -s -X POST http://0.0.0.0:8000/detokenize \
  -H "Content-Type: application/json" \
  -d '{"tokens": [1841, 603, 6479, 6044, 235336]}'
# Pooling: single text
curl -s -X POST http://0.0.0.0:8000/pooling \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","input":"What is artificial intelligence?"}'
# Pooling: batch texts
curl -s -X POST http://0.0.0.0:8000/pooling \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","input":["query one","query two","query three"]}'
# Pooling: image (chat format with base64)
curl -s -X POST http://0.0.0.0:8000/pooling \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","messages":[{"role":"user","content":[{"type":"image_url","image_url":{"url":"data:image/png;base64,<BASE64_STRING>"}}]}]}'
# Score: text vs text
curl -s -X POST http://0.0.0.0:8000/score \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","text_1":"What is machine learning?","text_2":["Machine learning is a subset of AI.","The weather is nice today."]}'
# Score: text vs image
curl -s -X POST http://0.0.0.0:8000/score \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","text_1":"a red square image","text_2":[{"content":[{"type":"image_url","image_url":{"url":"data:image/png;base64,<BASE64_STRING>"}}]}]}'
# Rerank
curl -s -X POST http://0.0.0.0:8000/rerank \
  -H "Content-Type: application/json" \
  -d '{"model":"vidore/colpali-v1.3-hf","query":"What is machine learning?","documents":["Machine learning is a branch of artificial intelligence.","The weather today is sunny and warm.","Deep learning uses neural networks with many layers.","Python is a popular programming language."]}'
```

## Test Result

| Endpoint | Status | Result |
|---|---|---|
| `/tokenize` | OK | `[1841, 603, 6479, 6044, 235336]` |
| `/detokenize` | OK | `"What is machine learning?"` |
| `/pooling` (text) | OK | 1 item, 5x128 |
| `/pooling` (batch) | OK | 3 items, 2x128 each |
| `/pooling` (image) | OK | 1 item, 1026x128 |
| `/score` (text vs text) | OK | ML=2.28, weather=0.85 |
| `/score` (text vs image) | OK | score=1.69 |
| `/rerank` | OK | ML(2.49) > DL(1.32) > Python(0.91) > weather(0.82) |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/colpali.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+    def _call_hf_processor(
+        self,
+        prompt: str,
+        mm_data: Mapping[str, object],
+        mm_kwargs: Mapping[str, object],
+        tok_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        if mm_data and not prompt:
+            prompt = "Describe the image."
```
Comment (DarkLight1337) [inline vllm/model_executor/models/colpali.py]: This should be the responsibility of the user, I think there is no need for us to provide a default prompt except through chat template.

File: vllm/model_executor/models/colpali.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            return hidden_states  # type: ignore
+
+        if self.custom_text_proj is not None:
+            proj_dtype = self.custom_text_proj.weight.dtype
+            if hidden_states.dtype != proj_dtype:
+                hidden_states = hidden_states.to(proj_dtype)
+            hidden_states = self.custom_text_proj(hidden_states)
+
+        # L2 normalize
+        return torch.nn.functional.normalize(hidden_states, p=2, dim=-1)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/colpali.py]: Normalization should be done by pooler

File: vllm/model_executor/models/colpali.py (hunk 1 above)
Comment (Kaonael) [reply vllm/model_executor/models/colpali.py]: Done

File: vllm/model_executor/models/colpali.py (hunk 2 above)
Comment (Kaonael) [reply vllm/model_executor/models/colpali.py]: Done

**Review by DarkLight1337 — COMMENTED** [review]
Thanks, LGTM. cc @noooop

**Review by noooop — APPROVED** [review]
